### PR TITLE
Get info about packages/pins from the IDE files.

### PR DESCRIPTION
**Pull request by yrabbit**
Signed-off-by: YRabbit <rabbit@yrabbit.cyou>

### CODE REVIEW

**Comment by yrabbit** [comment]
Well, we'll see how the CI goes. Should break on examples because nextpnr is old, but it will create bases.

File: apycula/tiled_fuzzer.py [hunk 1]
```diff
@@ -33,25 +33,22 @@
...
-        "package": "PG256",
-        "header": 0,
+        "package": "PBGA256",
         "device": "GW1N-4-PBGA256-6",
         "partnumber": "GW1N-LV4PG256C6/I5",
     },
     "GW1N-1": {
-        "package": "LQ144",
+        "package": "LQFP144",
         "header": 1, # stupid note in excel
```
Comment (pepijndevos) [inline apycula/tiled_fuzzer.py:52]: this header can go

File: apycula/chipdb.py [hunk 2]
```diff
@@ -234,51 +235,57 @@ def from_fse(fse):
...
+    for pkg_rec in pkgs.values():
+        pkg = pkg_rec[0]
+        if pkg in res:
+            continue
+        res[pkg] = pindef.get_pin_locs(device, pkg, pindef.VeryTrue)
+        res_bank_pins.update(pindef.get_bank_pins(device, pkg))
+    return (pkgs, res, res_bank_pins)
+
+# returns ({partnumber: (package, device)}, {pins}, {bank_pins})
+def xls_pinout(device):
```
Comment (pepijndevos) [inline apycula/chipdb.py:252]: maybe this shouldn't be called xls when it's not excel based ;)

**Comment by pepijndevos** [comment]
Exactly as predicted! So what's up with the nextpnr side? I mean, isn't the chipdb format the same?

File: apycula/gowin_bba.py [hunk 3]
```diff
@@ -147,6 +147,17 @@ def write_timing(b, timing):
     b.u32(len(timing))
     b.ref(blk)
 
+def write_partnumber_packages(b, db):
+    import ipdb; ipdb.set_trace()
```
Comment (pepijndevos) [inline apycula/gowin_bba.py:151]: uh breakpoints should go I think haha

Comment (yrabbit) [reply apycula/gowin_bba.py:151]: > Exactly as predicted! So what's up with the nextpnr side? I mean, isn't the chipdb format the same?

The structure has changed: a table with packages has been added.
You can put the next patch on nexpnr

https://github.com/yrabbit/nextpnr/commit/0e8a2999bda012175c1734abbf4efb76b515746c

**Comment by yrabbit** [comment]
The speed is determined from the partnumber, the match is stored in the database so the input regular expression is simplified and allows more models:)

Of course this is only friendly with the last commit for nextpnr:  https://github.com/yrabbit/nextpnr/commit/74b4f6972878d90c6d5c22450dae535559edbea2

**Comment by pepijndevos** [comment]
Looks good to me!

**Comment by pepijndevos** [comment]
Wait actually if this is an incompatible bba change it should bump the DB version

**Comment by pepijndevos** [comment]
https://github.com/YosysHQ/apicula/blob/b1dedc05ed4bf7ce622bd2e905c74867ac5331e6/apycula/gowin_bba.py#L186

File: examples/Makefile [hunk 4]
```diff
@@ -49,7 +49,7 @@ nanolcd-tangnano-synth.json: nanolcd/TOP.v nanolcd/VGAMod.v
 	$(YOSYS) -p "synth_gowin -json $@" $^
 
 %-tec0117-synth.json: %.v
-	$(YOSYS) -D LEDS_NR=8 -p "synth_gowin -json $@" $^
+	$(YOSYS) -D LEDS_NR=4 -p "synth_gowin -json $@" $^
```
Comment (pepijndevos) [inline examples/Makefile:52]: Why did the number of LEDs change? :|

**Pull request merged**